### PR TITLE
Fully implement the has_marker function

### PR DESCRIPTION
**Pull request by livia01px2019**
## What Changed?

`has_marker` fn implemented to: 
* Discover markers on the return of the controller
* Discover markers on nodes via the type of the node

## Why Does It Need To?

Self explanatory

## Checklist

- [x] Above description has been filled out so that upon quash merge we have a
  good record of what changed.
- [x] New functions, methods, types are documented. Old documentation is updated
  if necessary
- [x] Documentation in Notion has been updated
- [x] Tests for new behaviors are provided
  - [ ] New test suites (if any) ave been added to the CI tests (in
    `.github/workflows/rust.yml`) either as compiler test or integration test.
    *Or* justification for their omission from CI has been provided in this PR
    description.

### CODE REVIEW

File: crates/paralegal-policy/src/context.rs [hunk 1]
```diff
@@ -322,7 +322,18 @@ impl Context {
 
     /// Returns whether the given Node has the marker applied to it directly or via its type.
     pub fn has_marker(&self, marker: Marker, node: Node) -> bool {
-        // TODO: check the type as well
+        if let Some(src) = node.typ.as_data_source() {
+            if let Some(typ) = &self.desc.controllers[&node.ctrl_id].types.get(&src) {
```
Comment (JustusAdam) [inline crates/paralegal-policy/src/context.rs:326]: We should probably add a convenience function for getting the type(s) of a node.

File: crates/paralegal-policy/src/context.rs [hunk 2]
```diff
@@ -322,7 +322,18 @@ impl Context {
...
+                if typ.iter().any(|t| {
```
Comment (JustusAdam) [inline crates/paralegal-policy/src/context.rs:327]: This doesn't have to be an `if`, you can just return this expression. :)

File: crates/paralegal-policy/tests/test-crate/src/lib.rs [hunk 3]
```diff
@@ -16,9 +21,12 @@ fn cond(_f: Foo) -> bool {
 }
 
 #[paralegal_flow::analyze]
-fn controller(a: Foo, b: Foo) {
+#[paralegal_flow::marker{ ctrl, return }]
```
Comment (JustusAdam) [inline crates/paralegal-policy/tests/test-crate/src/lib.rs:24]: Nice!

File: crates/paralegal-flow/src/test_utils.rs [hunk 4]
```diff
@@ -135,7 +135,12 @@ where
     S: AsRef<std::ffi::OsStr>,
 {
     paralegal_flow_command(dir)
-        .args(["--abort-after-analysis", "--dump", "serialized-flow-graph"])
+        .args([
+            "--abort-after-analysis",
+            "--dump",
+            "serialized-flow-graph",
+            "--eager-local-markers",
```
Comment (JustusAdam) [inline crates/paralegal-flow/src/test_utils.rs:142]: Yeah I should really fix that.

**Review by JustusAdam — CHANGES_REQUESTED** [review]
Cool stuff!

File: crates/paralegal-policy/src/context.rs (hunk 1 above)
Comment (livia01px2019) [reply crates/paralegal-policy/src/context.rs:326]: What do you mean? So as_data_source etc functions on the node itself that just called it on the typ field?

File: crates/paralegal-policy/src/context.rs (hunk 2 above)
Comment (livia01px2019) [reply crates/paralegal-policy/src/context.rs:327]: We don't want to return if it's false!

File: crates/paralegal-flow/src/test_utils.rs (hunk 4 above)
Comment (livia01px2019) [reply crates/paralegal-flow/src/test_utils.rs:142]: I'll wait for #89 to be merged and rm this.

File: crates/paralegal-policy/src/context.rs (hunk 1 above)
Comment (JustusAdam) [reply crates/paralegal-policy/src/context.rs:326]: No I mean the essentially this expression `self.desc.controllers[&node.ctrl_id].types.get(&src)`, the Rust type if you will.

File: crates/paralegal-policy/src/context.rs (hunk 2 above)
Comment (JustusAdam) [reply crates/paralegal-policy/src/context.rs:327]: Oh sorry. I did not realize the context. I thought the function ended in 336.

**Review by JustusAdam — APPROVED** [review]

File: crates/paralegal-policy/src/context.rs (hunk 1 above)
Comment (livia01px2019) [reply crates/paralegal-policy/src/context.rs]: done

**Pull request merged**